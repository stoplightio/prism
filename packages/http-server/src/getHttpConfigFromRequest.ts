import { IHttpOperationConfig, IHttpRequest } from '@stoplight/prism-http';

export const getHttpConfigFromRequest = (req: IHttpRequest): Partial<IHttpOperationConfig> => {
  const httpOperationConfig: Partial<IHttpOperationConfig> = {};
  const query = req.url.query;

  if (!query) {
    return {};
  }

  const { __code, __dynamic, __example } = query;

  if (__code) {
    httpOperationConfig.code = typeof __code === 'string' ? __code : __code[0];
  }

  if (__dynamic) {
    httpOperationConfig.dynamic = __dynamic === 'true';
  }

  if (__example) {
    httpOperationConfig.exampleKey = typeof __example === 'string' ? __example : __example[0];
  }

  return httpOperationConfig;
};
