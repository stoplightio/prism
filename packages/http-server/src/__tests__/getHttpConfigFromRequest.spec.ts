import { getHttpConfigFromRequest } from '../getHttpConfigFromRequest';

describe('getHttpConfigFromRequest()', () => {
  describe('given no default config', () => {
    test('and no query should return my own default', () => {
      return expect(
        getHttpConfigFromRequest({
          method: 'get',
          url: { path: '/' },
        }),
      ).toMatchSnapshot();
    });
    test('and no matching query should return my own default', () => {
      return expect(
        getHttpConfigFromRequest({
          method: 'get',
          url: { path: '/', query: {} },
        }),
      ).toMatchSnapshot();
    });
    test('extracts code', () => {
      return expect(
        getHttpConfigFromRequest({
          method: 'get',
          url: { path: '/', query: { __code: '202' } },
        }),
      ).toMatchSnapshot();
    });
    test('extracts mediaType', () => {
      return expect(
        getHttpConfigFromRequest({
          method: 'get',
          url: { path: '/', query: { __contentType: 'application/json' } },
        }),
      ).toMatchSnapshot();
    });
    test('extracts example', () => {
      return expect(
        getHttpConfigFromRequest({
          method: 'get',
          url: { path: '/', query: { __example: 'bear' } },
        }),
      ).toMatchSnapshot();
    });
    test('extracts dynamic', () => {
      return expect(
        getHttpConfigFromRequest({
          method: 'get',
          url: { path: '/', query: { __dynamic: 'true' } },
        }),
      ).toMatchSnapshot();
    });
  });

  describe('given default config function', () => {
    test('and that function returns a mock should combine with query params', () => {
      const spy = jest.fn().mockReturnValue({
        mock: {
          exampleKey: 'key',
        },
      });
      return expect(
        getHttpConfigFromRequest(
          {
            method: 'get',
            url: { path: '/', query: { __code: '400' } },
          },
          spy,
        ),
      ).toMatchSnapshot();
    });
  });

  describe('given default config object', () => {
    test('that is boolean and no matching query params should return that config object', () => {
      return expect(
        getHttpConfigFromRequest(
          {
            method: 'get',
            url: { path: '/', query: {} },
          },
          { mock: false },
        ),
      ).toMatchSnapshot();
    });

    test('that is boolean and matching query should return that query', () => {
      return expect(
        getHttpConfigFromRequest(
          {
            method: 'get',
            url: { path: '/', query: { __code: '200', __example: 'bear' } },
          },
          { mock: false },
        ),
      ).toMatchSnapshot();
    });

    test('that is a map and matching query should return combined', () => {
      return expect(
        getHttpConfigFromRequest(
          {
            method: 'get',
            url: { path: '/', query: { __code: '200', __example: 'bear' } },
          },
          {
            mock: {
              dynamic: false,
              exampleKey: 'wolf',
              mediaType: 'plain/text',
            },
          },
        ),
      ).toMatchSnapshot();
    });
  });
});
