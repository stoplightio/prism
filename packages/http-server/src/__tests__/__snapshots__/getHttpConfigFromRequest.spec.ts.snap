// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getHttpConfigFromRequest() given default config function and that function returns a mock should combine with query params 1`] = `
Object {
  "mock": Object {
    "code": "400",
    "exampleKey": "key",
  },
}
`;

exports[`getHttpConfigFromRequest() given default config object that is a map and matching query should return combined 1`] = `
Object {
  "mock": Object {
    "code": "200",
    "dynamic": false,
    "exampleKey": "bear",
    "mediaType": "plain/text",
  },
}
`;

exports[`getHttpConfigFromRequest() given default config object that is boolean and matching query should return that query 1`] = `
Object {
  "mock": Object {
    "code": "200",
    "exampleKey": "bear",
  },
}
`;

exports[`getHttpConfigFromRequest() given default config object that is boolean and no matching query params should return that config object 1`] = `
Object {
  "mock": false,
}
`;

exports[`getHttpConfigFromRequest() given no default config and no matching query should return my own default 1`] = `Object {}`;

exports[`getHttpConfigFromRequest() given no default config and no query should return my own default 1`] = `Object {}`;

exports[`getHttpConfigFromRequest() given no default config extracts code 1`] = `
Object {
  "mock": Object {
    "code": "202",
  },
}
`;

exports[`getHttpConfigFromRequest() given no default config extracts dynamic 1`] = `
Object {
  "mock": Object {
    "dynamic": true,
  },
}
`;

exports[`getHttpConfigFromRequest() given no default config extracts example 1`] = `
Object {
  "mock": Object {
    "exampleKey": "bear",
  },
}
`;

exports[`getHttpConfigFromRequest() given no default config extracts mediaType 1`] = `
Object {
  "mock": Object {
    "mediaType": "application/json",
  },
}
`;
